File: 0540_Single_Element_in_a_Sorted_Array/main.cpp
// T = O(logn), S = O(1)
class Solution {
  public:
    int singleNonDuplicate(vector<int> &nums) {
        int n = nums.size();
        int l = 0, r = n - 1;
        while (l <= r) {
            int mid = (l + r) >> 1;
            if (mid - 1 >= 0 && nums[mid] == nums[mid - 1]) {
                mid--;
            } else if (mid + 1 < n && nums[mid] == nums[mid + 1]) {

            } else {
                return nums[mid];
            }
            if ((mid - l) % 2 == 1) {
                r = mid - 1;
            } else {
                l = mid + 2;
            }
        }
        return -1;
    }
};

File: 0540_Single_Element_in_a_Sorted_Array/golang/main.go
// T = O(logn), S = O(1)
func singleNonDuplicate(nums []int) int {
	if nums == nil || len(nums) == 0 {
		return -1
	}
	n := len(nums)
	l, r := 0, n-1
	for l <= r {
		mid := l + (r-l)/2
		if mid-1 >= 0 && nums[mid] == nums[mid-1] {
			mid--
		} else if mid+1 < n && nums[mid] == nums[mid+1] {

		} else {
			return nums[mid]
		}
		if mid&1 == 1 {
			r = mid - 1
		} else {
			l = mid + 2
		}
	}
	return -1
}